src/puj/push/program.clj

```diff
@@ -1,46 +1,67 @@
 (ns puj.push.program
   (:require [clojure.spec.alpha :as spec]
             [puj.push.unit :as u]
-            [puj.push.pushstate :as pushstate]))
+            [puj.push.type :as t]
+            [puj.push.config :as config]
+            [clojure.spec.alpha :as s]))
 
 
 (spec/def ::code (spec/and (spec/coll-of u/push-unit?) u/push-unit?))
-(spec/def ::io-scheme (spec/map-of keyword? keyword?))
+(spec/def ::io-scheme (spec/map-of keyword? ::t/type-name))
 (spec/def ::input-scheme ::io-scheme)
 (spec/def ::output-scheme ::io-scheme)
 
 
-(spec/def ::trigger
-  (spec/fspec :args [:old ::pushstate/state :new ::pushstate/state]
-              :ret boolean?))
-(spec/def ::revert-triggers (spec/coll-of ::trigger))
-(spec/def ::stop-triggers (spec/coll-of ::trigger))
+(spec/def ::signature
+  (spec/keys :req [::input-scheme ::output-scheme ::config/push-config]))
 
 
 (spec/def ::program
-  (spec/keys :req [::code ::input-scheme ::output-scheme]
-             :opt [::revert-triggers ::stop-triggers]))
+  (spec/keys :req [::code ::signature]))
 
 
 (defn make-program
-  [code inputs outputs & {:keys [revert-triggers stop-triggers]}]
-  {::code            code
-   ::input-scheme    inputs
-   ::output-scheme   outputs
-   ::revert-triggers revert-triggers
-   ::stop-triggers   stop-triggers})
+  ([code inputs outputs & {:keys [push-config]
+                           :or   {push-config (config/push-config)}}]
+   (make-program code
+                 {::input-scheme       inputs
+                  ::output-scheme      outputs
+                  ::config/push-config push-config}))
+  ([code signature]
```

Comment:
What does this function do? Could us a docstring.

Moreover, it looks like the entire body of the function is two maps. What does that do? I've never seen that before.

Reply:
Just added a docstring.

The first of the two maps is a [post-condition](https://clojure.org/reference/special_forms#_fn_name_param_condition_map_expr_2) that will raise and exception if the function's returned value is not a valid program.

The second map is just the map that gets returned. Such a tiny function isn't really needed because we could just create this map manually. The function lets us use the post-condition and it saves us from typing out the keys because the function uses positional arguments.